internal/winapi/process.go

```diff
@@ -1,10 +1,80 @@
 package winapi
 
+import (
+	"errors"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
 const PROCESS_ALL_ACCESS uint32 = 2097151
 
-// DWORD GetProcessImageFileNameW(
-//	HANDLE hProcess,
-//	LPWSTR lpImageFileName,
-//	DWORD  nSize
+const (
+	PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE = 0x20016
+	PROC_THREAD_ATTRIBUTE_JOB_LIST      = 0x2000D
+)
+
+type ProcThreadAttributeList struct {
+	_ [1]byte
+}
+
+// typedef struct _STARTUPINFOEXW {
+// 	STARTUPINFOW                 StartupInfo;
+// 	LPPROC_THREAD_ATTRIBUTE_LIST lpAttributeList;
+// } STARTUPINFOEXW, *LPSTARTUPINFOEXW;
+type StartupInfoEx struct {
+	// This is a recreation of the same binding from the stdlib. The x/sys/windows variant for whatever reason
+	// doesn't work when updating the list for the pseudo console attribute. It has the process immediately exit
+	// with exit code 0xc0000142 shortly after start.
+	windows.StartupInfo
+	ProcThreadAttributeList *ProcThreadAttributeList
+}
+
+// NewProcThreadAttributeList allocates a new ProcThreadAttributeList, with
+// the requested maximum number of attributes. This must be cleaned up by calling
+// DeleteProcThreadAttributeList.
+func NewProcThreadAttributeList(maxAttrCount uint32) (*ProcThreadAttributeList, error) {
+	var size uintptr
+	err := InitializeProcThreadAttributeList(nil, maxAttrCount, 0, &size)
+	if err != windows.ERROR_INSUFFICIENT_BUFFER {
+		if err == nil {
+			return nil, errors.New("unable to query buffer size from InitializeProcThreadAttributeList")
+		}
+		return nil, err
+	}
+	al := (*ProcThreadAttributeList)(unsafe.Pointer(&make([]byte, size)[0]))
+	err = InitializeProcThreadAttributeList(al, maxAttrCount, 0, &size)
+	if err != nil {
+		return nil, err
+	}
+	return al, nil
+}
+
+// BOOL InitializeProcThreadAttributeList(
+// 	[out, optional] LPPROC_THREAD_ATTRIBUTE_LIST lpAttributeList,
+// 	[in]            DWORD                        dwAttributeCount,
+// 					DWORD                        dwFlags,
+// 	[in, out]       PSIZE_T                      lpSize
 // );
-//sys GetProcessImageFileName(hProcess windows.Handle, imageFileName *uint16, nSize uint32) (size uint32, err error) = kernel32.GetProcessImageFileNameW
+//
+//sys InitializeProcThreadAttributeList(lpAttributeList *ProcThreadAttributeList, dwAttributeCount uint32, dwFlags uint32, lpSize *uintptr) (err error) = kernel32.InitializeProcThreadAttributeList
```

Comment:
does this need to be exported?

Reply:
I suppose no, as we only use it in the helper we've made.

Reply:
Unexported